<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели-и-задачи" data-toc-modified-id="Цели-и-задачи-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели и задачи</a></span></li><li><span><a href="#Импорт-данных" data-toc-modified-id="Импорт-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Импорт данных</a></span></li><li><span><a href="#Выполнение-заданий" data-toc-modified-id="Выполнение-заданий-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выполнение заданий</a></span><ul class="toc-item"><li><span><a href="#Количество-книг-вышедшие-после-1-января-2000-года" data-toc-modified-id="Количество-книг-вышедшие-после-1-января-2000-года-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Количество книг вышедшие после 1 января 2000 года</a></span></li><li><span><a href="#Количество-обзор-и-средняя-оценка-для-каждой-книги" data-toc-modified-id="Количество-обзор-и-средняя-оценка-для-каждой-книги-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Количество обзор и средняя оценка для каждой книги</a></span></li><li><span><a href="#Издательство,которое-выпустило-наибольшое-количество-книг" data-toc-modified-id="Издательство,которое-выпустило-наибольшое-количество-книг-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Издательство,которое выпустило наибольшое количество книг</a></span></li><li><span><a href="#Автор-с-самой-высокой-оценкой" data-toc-modified-id="Автор-с-самой-высокой-оценкой-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Автор с самой высокой оценкой</a></span></li><li><span><a href="#Среднее-количество-оценок-от-каждого-пользователя" data-toc-modified-id="Среднее-количество-оценок-от-каждого-пользователя-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Среднее количество оценок от каждого пользователя</a></span></li><li><span><a href="#Вывод" data-toc-modified-id="Вывод-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Вывод</a></span></li></ul></li></ul></div>

# Цели и задачи

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Импорт данных

In [1]:
import pandas as pd 
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
books = pd.io.sql.read_sql('select * from books', con = engine) 
display(books)
display(books.info())


book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


None

In [4]:
authors = pd.io.sql.read_sql('select * from authors', con = engine) 
display(authors)
display(authors.info())

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

In [5]:
publishers = pd.io.sql.read_sql('select * from publishers', con = engine) 
display(publishers)
display(publishers.info())

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


None

In [6]:
ratings = pd.io.sql.read_sql('select * from ratings', con = engine) 
display(ratings)
display(ratings.info())

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

In [7]:
reviews = pd.io.sql.read_sql('select * from reviews', con = engine) 
display(reviews)
display(reviews.info())

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

# Выполнение заданий

In [8]:
def sql(request):
    display(pd.io.sql.read_sql(request, con=engine))

## Количество книг вышедшие после 1 января 2000 года

In [9]:
pd.io.sql.read_sql("SELECT count(book_id) \
                   FROM BOOKS \
                   where publication_date > '2000-01-01'" , con = engine) 

count
0    819

Количество книг, которые вышли после 1 января 2000 года

## Количество обзор и средняя оценка для каждой книги

In [10]:
sql('Select count(distinct(book_id)) from reviews')

count
0    994

In [11]:
sql('SELECT * FROM books WHERE title IN (SELECT title FROM books GROUP BY title HAVING COUNT(*) > 1) ORDER BY title')

book_id  author_id                title  num_pages publication_date  \
0      426         39  Memoirs of a Geisha        434       2005-11-15   
1      427         39  Memoirs of a Geisha        503       2005-11-22   

   publisher_id  
0           241  
1           311

In [12]:
pd.io.sql.read_sql("SELECT b.title, ROUND(AVG(RT.RATING),1), COUNT(distinct(review_id)) as review \
from books b JOIN RATINGS RT ON B.BOOK_ID = RT.BOOK_ID \
left JOIN reviews RV ON b.BOOK_ID = RV.BOOK_ID GROUP BY TITLE order by round desc, review desc " , con = engine) 

title  round  review
0                    A Dirty Job (Grim Reaper  #1)    5.0       4
1     Moneyball: The Art of Winning an Unfair Game    5.0       3
2          School's Out—Forever (Maximum Ride  #2)    5.0       3
3    Arrows of the Queen (Heralds of Valdemar  #1)    5.0       2
4                                    Evening Class    5.0       2
..                                             ...    ...     ...
994                         The Kitchen God's Wife    2.3       3
995                                  Drowning Ruth    2.0       3
996                                          Junky    2.0       2
997              His Excellency: George Washington    2.0       2
998                           Harvesting the Heart    1.5       2

[999 rows x 3 columns]

Количество обзоров и средняя оценка для каждой книги. Как мы видим не на все книги есть обзоры и оценки

## Издательство,которое выпустило наибольшое количество книг

In [13]:
pd.io.sql.read_sql("SELECT count(book_id), \
publisher from books b JOIN publishers pb ON b.publisher_id = pb.publisher_id \
where num_pages>50 group by publisher order by count desc limit 1" , con = engine) 

count      publisher
0     42  Penguin Books

Издательство, которые выпустило наибольшее количество книг толщиной более 50 страниц

## Автор с самой высокой оценкой

In [14]:
pd.io.sql.read_sql('SELECT A.AUTHOR, ROUND(AVG(RT.RATING),1) rating FROM BOOKS B JOIN AUTHORS \
A ON B.AUTHOR_ID = A.AUTHOR_ID JOIN RATINGS RT ON B.BOOK_ID = RT.BOOK_ID \
GROUP BY A.AUTHOR HAVING COUNT(RT.RATING_ID)>50 ORDER BY rating desc limit 1',con = engine)

author  rating
0  J.K. Rowling/Mary GrandPré     4.3

Автор с самой высокой средней оценкой

## Среднее количество оценок от каждого пользователя

In [15]:
pd.io.sql.read_sql('SELECT ROUND(AVG(COUNT),1) \
FROM (SELECT COUNT(REVIEW_ID) \
FROM (SELECT USERNAME FROM RATINGS GROUP BY USERNAME HAVING COUNT(RATING_ID)>50) \
RT JOIN REVIEWS RV ON RT.USERNAME = RV.USERNAME GROUP BY RV.USERNAME)T',con = engine)

round
0   24.3

Среднее количество оценок

## Вывод

* Всего с начала 2000 года вышло 819 книг
* Издательство, которое выпустило наибольшее количество книг - Penguin Books
* Автор с самой высокой оценкой - J.K. Rowling/Mary GrandPré при оценке 4.3
* Среднее количество оценок от каждого пользователя - 24